### PR TITLE
fix: allow for feedstock_root to not be set

### PR DESCRIPTION
**Pull request by beckermr**
This PR fixes the rerendering bug where FEEDSTOCK_ROOT is not set and smithy chokes.

### CODE REVIEW

**Comment by beckermr** [comment]
@conda-forge-admin rerender

**Comment by leofang** [comment]
Thanks, Matt! We've been seeing weird rerendering errors since this morning https://github.com/conda-forge/openmpi-feedstock/pull/175#issuecomment-2374430522, I assume it's the same issue?

**Comment by beckermr** [comment]
yes it appears to be the same issue.

**Comment by jakirkham** [comment]
Well there are two errors we are seeing:

1. `FEEDSTOCK_ROOT` in this recipe causes re-rendering issues
2. `CONDA_BUILD_SYSROOT` not being set, which causes various odd build time issues ( https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/349#issuecomment-2375439341 )

Matt is working around the first issue for this feedstock specifically

We are still investigating what causes either error (or how to fix them)

**Comment by jakirkham** [comment]
Ok think I understand 1. It is due to a conda-build change in 24.7.0. Namely this one: https://github.com/conda/conda-build/pull/5371

Was able to re-render with conda-build 24.5 as show in PR: https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/352

Think we need to add some handling for re-rendering like this PR to address it

At least the cause of the first error is now clearer

File: recipe/meta.yaml [hunk 1]
```diff
@@ -21,9 +21,9 @@ build:
   skip: True
   {% endif %}
   script:
-    - cp {{ os.environ["FEEDSTOCK_ROOT"] }}/LICENSE.txt ${RECIPE_DIR}/LICENSE.txt            # [linux]
-    - cp ${RECIPE_DIR}/../LICENSE.txt ${RECIPE_DIR}/LICENSE.txt                              # [osx]
-    - COPY "%RECIPE_DIR%\\..\\LICENSE.txt" "%RECIPE_DIR%\\LICENSE.txt"                       # [win]
+    - cp {{ os.environ.get("FEEDSTOCK_ROOT", "failme") }}/LICENSE.txt ${RECIPE_DIR}/LICENSE.txt  # [linux]
```
Comment (jakirkham) [inline recipe/meta.yaml:24]: We might consider adding a guard here. Also we could flesh out the error message a bit

```suggestion
    {% if "FEEDSTOCK_ROOT" in os.environ %}
    - cp {{ os.environ["FEEDSTOCK_ROOT"] }}/LICENSE.txt ${RECIPE_DIR}/LICENSE.txt  # [linux]
    {% else %}
    - echo '${FEEDSTOCK_ROOT} is undefined. Cannot copy license file'
    - exit 1
    {% endif %}
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -21,9 +21,14 @@ build:
...
-    - cp ${RECIPE_DIR}/../LICENSE.txt ${RECIPE_DIR}/LICENSE.txt                              # [osx]
-    - COPY "%RECIPE_DIR%\\..\\LICENSE.txt" "%RECIPE_DIR%\\LICENSE.txt"                       # [win]
+    {% if "FEEDSTOCK_ROOT" in os.environ %}
+    - cp {{ os.environ["FEEDSTOCK_ROOT"] }}/LICENSE.txt ${RECIPE_DIR}/LICENSE.txt  # [linux]
+    {% else %}
+    - echo '${FEEDSTOCK_ROOT} is undefined. Cannot copy license file'
+    - exit 1
+    {% endif %}
+    - cp ${RECIPE_DIR}/../LICENSE.txt ${RECIPE_DIR}/LICENSE.txt                                  # [osx]
+    - COPY "%RECIPE_DIR%\\..\\LICENSE.txt" "%RECIPE_DIR%\\LICENSE.txt"                           # [win]
```
Comment (jakirkham) [inline recipe/meta.yaml:31]: Forgot we need more Linux selectors

```suggestion
    {% if "FEEDSTOCK_ROOT" in os.environ %}                                                  # [linux]
    - cp {{ os.environ["FEEDSTOCK_ROOT"] }}/LICENSE.txt ${RECIPE_DIR}/LICENSE.txt            # [linux]
    {% else %}                                                                               # [linux]
    - echo '${FEEDSTOCK_ROOT} is undefined. Cannot copy license file'                        # [linux]
    - exit 1                                                                                 # [linux]
    {% endif %}                                                                              # [linux]
    - cp ${RECIPE_DIR}/../LICENSE.txt ${RECIPE_DIR}/LICENSE.txt                              # [osx]
    - COPY "%RECIPE_DIR%\\..\\LICENSE.txt" "%RECIPE_DIR%\\LICENSE.txt"                       # [win]
```

**Comment by jakirkham** [comment]
@conda-forge-admin , please re-render

(just to reconfirm it works)

**Comment by jakirkham** [comment]
Looks like re-rendering simply renamed variant files. So nothing really changed

Main thing is we are able to re-render again

**Comment by beckermr** [comment]
Let's wait to merge and see if we can release a bugfix all at once for both issues.

**Comment by beckermr** [comment]
@conda-forge-admin restart ci

**Comment by beckermr** [comment]
@conda-forge-admin rerender

**Comment by leofang** [comment]
Nice, what change fixed the rerender error?

**Comment by beckermr** [comment]
We need to define some vars at the top if they are undefined.

**Pull request merged**